File: thbigmatome-front/src/components/squad/__tests__/SquadTextSettings.spec.ts
import { describe, it, expect, vi, beforeEach } from 'vitest'
import { mount, flushPromises } from '@vue/test-utils'
import { createVuetify } from 'vuetify'
import * as components from 'vuetify/components'
import * as directives from 'vuetify/directives'
import { createI18n } from 'vue-i18n'
import SquadTextSettings from '../SquadTextSettings.vue'

vi.mock('axios', () => ({
  default: {
    get: vi.fn(),
    put: vi.fn(),
    defaults: { baseURL: '', withCredentials: false, headers: { common: {} } },
    interceptors: { request: { use: vi.fn() }, response: { use: vi.fn() } },
  },
}))

import axios from 'axios'

const vuetify = createVuetify({ components, directives })
const i18n = createI18n({
  legacy: false,
  locale: 'ja',
  messages: {
    ja: {
      squadTextSettings: {
        title: 'スカッドテキスト書式設定',
        positionFormat: 'ポジション表記',
        positionFormatEnglish: '英語略称',
        positionFormatJapanese: '漢字略称',
        handednessFormat: '投打表記',
        handednessFormatAlphabet: 'アルファベット',
        handednessFormatKanji: '漢字',
        dateFormat: '登板履歴日付',
        dateFormatAbsolute: '絶対日付',
        dateFormatRelative: '相対日付',
        sectionHeaderFormat: 'セクションヘッダー',
        sectionHeaderFormatBracket: '括弧付き',
        sectionHeaderFormatNone: '無印',
        showNumberPrefix: '背番号接頭辞',
        show: '表示する',
        hide: '表示しない',
        battingStatsItems: '打者成績項目',
        pitchingStatsItems: '投手成績項目',
        battingStats: {
          avg: '打率',
          hr: '本塁打',
          rbi: '打点',
          sb: '盗塁',
          obp: '出塁率',
          ops: 'OPS',
        },
        pitchingStats: {
          w_l: '勝敗',
          games: '登板数',
          era: '防御率',
          so: '奪三振',
          ip: '投球回',
          hold: 'ホールド',
          save: 'セーブ',
        },
        save: '保存',
        saved: '保存しました',
        saveError: '保存に失敗しました',
        loadError: '設定の読み込みに失敗しました',
      },
    },
  },
})

const mockSettingData = {
  id: 1,
  team_id: 10,
  position_format: 'english',
  handedness_format: 'alphabet',
  date_format: 'absolute',
  section_header_format: 'bracket',
  show_number_prefix: true,
  batting_stats_config: { avg: true, hr: true, rbi: true, sb: false, obp: false, ops: false },
  pitching_stats_config: {
    w_l: true,
    games: true,
    era: true,
    so: true,
    ip: true,
    hold: false,
    save: false,
  },
  updated_at: '2026-03-07T01:00:00.000Z',
}

function mountComponent() {
  return mount(SquadTextSettings, {
    props: { teamId: 10 },
    global: {
      plugins: [vuetify, i18n],
    },
  })
}

describe('SquadTextSettings', () => {
  beforeEach(() => {
    vi.clearAllMocks()
  })

  describe('初期ロード', () => {
    it('マウント時にGETリクエストを送信する', async () => {
      vi.mocked(axios.get).mockResolvedValueOnce({ data: mockSettingData })
      mountComponent()
      expect(axios.get).toHaveBeenCalledWith('/teams/10/squad_text_settings')
    })

    it('タイトルが表示される', async () => {
      vi.mocked(axios.get).mockResolvedValueOnce({ data: mockSettingData })
      const wrapper = mountComponent()
      await flushPromises()
      expect(wrapper.text()).toContain('スカッドテキスト書式設定')
    })

    it('ロード後にポジション表記セクションが表示される', async () => {
      vi.mocked(axios.get).mockResolvedValueOnce({ data: mockSettingData })
      const wrapper = mountComponent()
      await flushPromises()
      expect(wrapper.text()).toContain('ポジション表記')
    })

    it('ロード後に投打表記セクションが表示される', async () => {
      vi.mocked(axios.get).mockResolvedValueOnce({ data: mockSettingData })
      const wrapper = mountComponent()
      await flushPromises()
      expect(wrapper.text()).toContain('投打表記')
    })

    it('ロード後に打者成績項目セクションが表示される', async () => {
      vi.mocked(axios.get).mockResolvedValueOnce({ data: mockSettingData })
      const wrapper = mountComponent()
      await flushPromises()
      expect(wrapper.text()).toContain('打者成績項目')
    })

    it('ロード後に投手成績項目セクションが表示される', async () => {
      vi.mocked(axios.get).mockResolvedValueOnce({ data: mockSettingData })
      const wrapper = mountComponent()
      await flushPromises()
      expect(wrapper.text()).toContain('投手成績項目')
    })

    it('ロード後に保存ボタンが表示される', async () => {
      vi.mocked(axios.get).mockResolvedValueOnce({ data: mockSettingData })
      const wrapper = mountComponent()
      await flushPromises()
      expect(wrapper.text()).toContain('保存')
    })
  })

  describe('設定の表示', () => {
    it('position_formatがenglishの場合、英語略称ラジオが選択される', async () => {
      vi.mocked(axios.get).mockResolvedValueOnce({
        data: { ...mockSettingData, position_format: 'english' },
      })
      const wrapper = mountComponent()
      await flushPromises()
      expect(wrapper.text()).toContain('英語略称')
    })

    it('各成績項目ラベルが表示される', async () => {
      vi.mocked(axios.get).mockResolvedValueOnce({ data: mockSettingData })
      const wrapper = mountComponent()
      await flushPromises()
      expect(wrapper.text()).toContain('打率')
      expect(wrapper.text()).toContain('本塁打')
      expect(wrapper.text()).toContain('打点')
      expect(wrapper.text()).toContain('防御率')
      expect(wrapper.text()).toContain('奪三振')
    })
  })

  describe('保存', () => {
    it('保存ボタンクリック時にPUTリクエストを送信する', async () => {
      vi.mocked(axios.get).mockResolvedValueOnce({ data: mockSettingData })
      vi.mocked(axios.put).mockResolvedValueOnce({ data: mockSettingData })
      const wrapper = mountComponent()
      await flushPromises()
      await wrapper.find('button').trigger('click')
      await flushPromises()
      expect(axios.put).toHaveBeenCalledWith(
        '/teams/10/squad_text_settings',
        expect.objectContaining({
          squad_text_setting: expect.objectContaining({
            position_format: 'english',
            handedness_format: 'alphabet',
          }),
        }),
      )
    })

    it('保存失敗時にsnackbarがエラー状態になる', async () => {
      vi.mocked(axios.get).mockResolvedValueOnce({ data: mockSettingData })
      vi.mocked(axios.put).mockRejectedValueOnce(new Error('Network Error'))
      const wrapper = mountComponent()
      await flushPromises()
      await wrapper.find('button').trigger('click')
      await flushPromises()
      const vm = wrapper.vm as { snackbar: { show: boolean; message: string; color: string } }
      expect(vm.snackbar.show).toBe(true)
      expect(vm.snackbar.color).toBe('error')
      expect(vm.snackbar.message).toContain('失敗')
    })
  })

  describe('GETエラー', () => {
    it('ロード失敗時にsnackbarがエラー状態になる', async () => {
      vi.mocked(axios.get).mockRejectedValueOnce(new Error('Network Error'))
      const wrapper = mountComponent()
      await flushPromises()
      const vm = wrapper.vm as { snackbar: { show: boolean; message: string; color: string } }
      expect(vm.snackbar.show).toBe(true)
      expect(vm.snackbar.color).toBe('error')
      expect(vm.snackbar.message).toContain('失敗')
    })
  })
})

File: thbigmatome/spec/requests/api/v1/squad_text_settings_spec.rb
require "rails_helper"

RSpec.describe "Api::V1::SquadTextSettings", type: :request do
  let(:team) { create(:team) }

  def squad_text_settings_url
    "/api/v1/teams/#{team.id}/squad_text_settings"
  end

  describe "認証なしアクセス" do
    it "GET /squad_text_settings → 401" do
      get squad_text_settings_url
      expect(response).to have_http_status(:unauthorized)
    end

    it "PUT /squad_text_settings → 401" do
      put squad_text_settings_url, params: { squad_text_setting: { position_format: "japanese" } }, as: :json
      expect(response).to have_http_status(:unauthorized)
    end
  end

  describe "GET /api/v1/teams/:team_id/squad_text_settings" do
    include_context "authenticated user"
    before { team.update!(user: user) }

    context "設定が存在する場合" do
      let!(:setting) { create(:squad_text_setting, team: team, position_format: "japanese") }

      it "200を返す" do
        get squad_text_settings_url
        expect(response).to have_http_status(:ok)
      end

      it "既存の設定を返す" do
        get squad_text_settings_url
        json = JSON.parse(response.body)
        expect(json["id"]).to eq(setting.id)
        expect(json["position_format"]).to eq("japanese")
        expect(json["team_id"]).to eq(team.id)
        expect(json).to have_key("batting_stats_config")
        expect(json).to have_key("pitching_stats_config")
        expect(json).to have_key("updated_at")
      end
    end

    context "設定が存在しない場合（初回アクセス）" do
      it "200を返す（404しない）" do
        get squad_text_settings_url
        expect(response).to have_http_status(:ok)
      end

      it "デフォルト値で新規作成してレスポンスを返す" do
        expect {
          get squad_text_settings_url
        }.to change(SquadTextSetting, :count).by(1)

        json = JSON.parse(response.body)
        expect(json["position_format"]).to eq("english")
        expect(json["handedness_format"]).to eq("alphabet")
        expect(json["date_format"]).to eq("absolute")
        expect(json["section_header_format"]).to eq("bracket")
        expect(json["show_number_prefix"]).to eq(true)
      end

      it "batting_stats_configにデフォルト値が含まれる" do
        get squad_text_settings_url
        json = JSON.parse(response.body)
        config = json["batting_stats_config"]
        expect(config["avg"]).to eq(true)
        expect(config["hr"]).to eq(true)
        expect(config["rbi"]).to eq(true)
      end

      it "pitching_stats_configにデフォルト値が含まれる" do
        get squad_text_settings_url
        json = JSON.parse(response.body)
        config = json["pitching_stats_config"]
        expect(config["era"]).to eq(true)
        expect(config["so"]).to eq(true)
      end
    end
  end

  describe "PUT /api/v1/teams/:team_id/squad_text_settings" do
    include_context "authenticated user"
    before { team.update!(user: user) }

    context "設定が存在する場合（update）" do
      let!(:setting) { create(:squad_text_setting, team: team) }

      it "200を返す" do
        put squad_text_settings_url,
            params: { squad_text_setting: { position_format: "japanese" } },
            as: :json
        expect(response).to have_http_status(:ok)
      end

      it "レコード数が増えない" do
        expect {
          put squad_text_settings_url,
              params: { squad_text_setting: { position_format: "japanese" } },
              as: :json
        }.not_to change(SquadTextSetting, :count)
      end

      it "設定が更新される" do
        put squad_text_settings_url,
            params: { squad_text_setting: { position_format: "japanese", handedness_format: "kanji" } },
            as: :json
        json = JSON.parse(response.body)
        expect(json["position_format"]).to eq("japanese")
        expect(json["handedness_format"]).to eq("kanji")
      end
    end

    context "設定が存在しない場合（create）" do
      it "200を返す" do
        put squad_text_settings_url,
            params: { squad_text_setting: { position_format: "japanese" } },
            as: :json
        expect(response).to have_http_status(:ok)
      end

      it "新規レコードを作成する" do
        expect {
          put squad_text_settings_url,
              params: { squad_text_setting: { position_format: "japanese" } },
              as: :json
        }.to change(SquadTextSetting, :count).by(1)
      end

      it "指定した設定値で作成される" do
        put squad_text_settings_url,
            params: { squad_text_setting: { position_format: "japanese" } },
            as: :json
        json = JSON.parse(response.body)
        expect(json["position_format"]).to eq("japanese")
      end
    end

    context "batting_stats_config / pitching_stats_config の更新" do
      let!(:setting) { create(:squad_text_setting, team: team) }

      it "batting_stats_configを更新できる" do
        put squad_text_settings_url,
            params: {
              squad_text_setting: {
                batting_stats_config: { "avg" => true, "hr" => true, "rbi" => true, "sb" => true, "obp" => false, "ops" => true, "ab_h" => false }
              }
            },
            as: :json
        json = JSON.parse(response.body)
        expect(json["batting_stats_config"]["sb"]).to eq(true)
        expect(json["batting_stats_config"]["ops"]).to eq(true)
      end

      it "pitching_stats_configを更新できる" do
        put squad_text_settings_url,
            params: {
              squad_text_setting: {
                pitching_stats_config: { "w_l" => true, "games" => true, "era" => true, "so" => true, "ip" => true, "hold" => true, "save" => true }
              }
            },
            as: :json
        json = JSON.parse(response.body)
        expect(json["pitching_stats_config"]["hold"]).to eq(true)
        expect(json["pitching_stats_config"]["save"]).to eq(true)
      end
    end

    context "show_number_prefixの更新" do
      let!(:setting) { create(:squad_text_setting, team: team, show_number_prefix: true) }

      it "falseに更新できる" do
        put squad_text_settings_url,
            params: { squad_text_setting: { show_number_prefix: false } },
            as: :json
        json = JSON.parse(response.body)
        expect(json["show_number_prefix"]).to eq(false)
      end
    end
  end
end
